apps/react/src/features/genres/components/GenresList/GenresList.tsx

```diff
@@ -1,23 +1,95 @@
-import { Genre } from '@js-camp/core/models/genre';
-import { List } from '@mui/material';
-
-import { GenreCard } from '../GenreCard';
+import { FC, memo, useCallback, useEffect, useRef, useState } from 'react';
+import { useSearchParams } from 'react-router-dom';
+import { List, Typography } from '@mui/material';
+import CircularProgress from '@mui/material/CircularProgress';
+import { useAppDispatch, useAppSelector } from '@js-camp/react/store';
+import {
+	selectAreGenresLoading,
+	selectGenres,
+	selectGenresError,
+	selectGenresHasMore,
+} from '@js-camp/react/store/genre/selectors';
+import { DEFAULT_PAGE_NUMBER, DEFAULT_PAGE_SIZE } from '@js-camp/react/config';
+import { GenreCard } from '@js-camp/react/features/genres/components/GenreCard/GenreCard';
+import { fetchGenres } from '@js-camp/react/store/genre/dispatchers';
+import { getQueryParams } from '@js-camp/react/utils/getQueryParams';
 
 import styles from './GenresList.module.css';
 
-type GenreListProps = {
+/** Renders a list of genres. */
+export const GenresListComponent: FC = () => {
+	const [searchParams] = useSearchParams();
+	const dispatch = useAppDispatch();
+	const isLoading = useAppSelector(selectAreGenresLoading);
+	const error = useAppSelector(selectGenresError);
+	const genres = useAppSelector(selectGenres);
+	const hasMore = useAppSelector(selectGenresHasMore);
+	const [pageNumber, setPageNumber] = useState(DEFAULT_PAGE_NUMBER);
+	const observer = useRef<IntersectionObserver>();
+	const wrapperElementRef = useRef<HTMLElement>(null);
+	const noGenres = !isLoading && !error && !genres.length;
 
-	/** An array of genres. */
-	genres: readonly Genre[];
-};
+	const lastGenreElementRef = useCallback((node: HTMLLIElement | null) => {
+		observer.current?.disconnect();
 
-/** Renders a list of genres. */
-export const GenresListComponent: FC<GenreListProps> = ({ genres }: {genres: readonly Genre[];}) => (
-	<List className={styles['genres-list']} disablePadding>
-		{genres.map(genre => <GenreCard key={genre.id} genre={genre} />)}
-	</List>
-);
+		if (node) {
+			observer.current = new IntersectionObserver(entries => {
+				if (entries[0].isIntersecting && hasMore) {
+					setPageNumber(prev => prev + 1);
+					observer.current?.disconnect();
+				}
+			});
+
+			observer.current.observe(node);
+		}
+	}, [hasMore]);
+
+	const { search, sort, filter } = getQueryParams(searchParams);
+
+	const dispatchGenres = useCallback(
+		(newPageNumber: number) => dispatch(fetchGenres({
+			search,
+			sort,
+			filter,
+			pageNumber: newPageNumber,
+			pageSize: DEFAULT_PAGE_SIZE,
+		})),
+		[search, sort, filter],
+	);
+
+	useEffect(() => {
+		wrapperElementRef.current?.scrollTo({ top: 0, behavior: 'instant' });
+		setPageNumber(DEFAULT_PAGE_NUMBER);
+		dispatchGenres(DEFAULT_PAGE_NUMBER);
+	}, [searchParams]);
+
+	useEffect(() => {
+		dispatchGenres(pageNumber);
+	}, [pageNumber]);
+
+	return <section ref={wrapperElementRef} className={styles['genres-list-wrapper']}>
```

Comment:
Why don't you use `Box` component? It can also accept refs and you could pass a props called "variant" and specify it as "section".

This way you won't have a mix of MUI components and html ones  in your codebase :)

```suggestion
  <Box variant="section" ref={wrapperElementRef}>
     // other elements 
  </Box>
```